content: add lesson 22 strings

johnnygossdev opened the pull request:
This PR adds lesson 22: Strings. It's a relatively short lesson which looks are what strings are and why we use them.

It has a few practices which use strings with other learnt concepts like arrays.

- [x] Write lesson
- [x] Add practices

closes #308

![image](https://user-images.githubusercontent.com/3285214/154840872-c731a3e4-e088-4295-8ad8-dfe6fc4311f9.png)


https://user-images.githubusercontent.com/3285214/155763313-4dbe2fcb-d002-450d-a5f0-7bd38e2a88dc.mp4

johnnygossdev commented:
This lesson is ready for review @NathanLovato, but I'm unable to merge `main` into this branch and resolve conflicts due to filenames which aren't compatable with Windows. Jad mentioned the `::` was the issue.

```
Error: Command failed: C:\Program Files\Git\cmd\git.exe -c core.longpaths=true -c core.quotepath=false -c color.ui=false checkout main --
error: invalid path 'course/lesson-19-creating-arrays/moving-turtle/MovingTurtle.live-editor/scripts/MovingTurtle.tscn::1.tres'
```

johnnygossdev commented:
Also I'd like to catch this error because missing `""` could be pretty common with strings.

![image](https://user-images.githubusercontent.com/3285214/154854470-3bad40d2-2f79-4767-8aab-5306f13c5133.png)

But the app crashes instead.

![image](https://user-images.githubusercontent.com/3285214/154854492-9d853e2a-add4-4092-a6b7-61daf5463739.png)

johnnygossdev commented:
As a final note, it'd be good to ceate punching animations and swap the default animation with that, but I didn't have time to create them.

NathanLovato commented on course/lesson-22-strings/lesson.tres:
```diff
@@ -0,0 +1,198 @@
+content_id = "res://course/lesson-22-strings/content-DpPwQDwf.tres"
+title = "What are strings"
+type = 0
+text = "A string is a variable type which holds text instead of numbers. When we use strings, we use quotation marks (\"\") to differentiate them from other variable types, or method names for example.
```
You shouldn't talk about variable type, at this stage, variables are dynamic. Also, String itself is a _value_ type, you don't need a variable to hold it.

NathanLovato commented on course/lesson-22-strings/lesson.tres:
```diff
@@ -0,0 +1,198 @@
+content_id = "res://course/lesson-22-strings/content-jotxF5HU.tres"
+title = ""
+type = 0
+text = "Throughout this course, we've mostly stored numbers in variables. But what if we wanted to store a player's name? Using only numbers for names isn't very user-friendly at all, as 70801 here will tell you."
```
Actually, this first example, which gives the robot a number as a name, uses a String. I'm unsure we should use that as a result, because as a dev you need to be really clear with types, their differences, and when you need to convert them (like here, you'd convert a number to a string to display it in a label)

NathanLovato commented on course/lesson-22-strings/lesson.tres:
```diff
@@ -0,0 +1,198 @@
+content_id = "res://course/lesson-22-strings/content-5tT6n1Tx.tres"
+title = ""
+type = 0
+text = "Strings are like an array of characters. In fact, we can even loop through the characters of a string in the same way."
```
Not _like_, they are arrays of characters. Just in GDScript, it's happening under the hood as there's no exposed character type.

NathanLovato commented on course/lesson-22-strings/lesson.tres:
```diff
@@ -0,0 +1,198 @@
+type = 0
+text = "Every piece of text you see in this app is a string that Godot is displaying for us.
+
+Much like how [code]Vector2[/code] variables make calculations easier, string variables come with lots of helper methods and tricks we can use.
```
We've used the term function throughout the app so far. We're not gonna cover OOP or objects so we can't really talk about methods and properties, which are OO concepts

NathanLovato commented:
Just left you a few comments for future updates

> As a final note, it'd be good to ceate punching animations and swap the default animation with that, but I didn't have time to create them.

There's no time limit on this. Can you add them? The combo practice doesn't feel like a combo without them. If not let me know I'll take care of those

johnnygossdev commented:
Great, thanks. Sure, I can add the animations and perhaps remove the first visual with the number name. I thought it might be misleading when I added it, but wasn't sure how else to add a visual to it. Perhaps it's not needed.

NathanLovato commented:
I think it's fine with mostly code/print examples. It's a short lesson so it's not so long it'd get boring for the reader.

NathanLovato commented:
Hey Johnny, is this good to test and merge now?

johnnygossdev commented:
Hello!

I'd say so. I added different animations and updated the practice. There's
perhaps removing the first potentially confusing visual in the lesson with
the number. Otherwise good to go

On Sun, 27 Feb 2022, 09:17 Nathan Lovato, ***@***.***> wrote:

> Hey Johnny, is this good to test and merge now?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/GDQuest/learn-gdscript/pull/345#issuecomment-1053408815>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZCBXUA3DLQOIC7RS7VBKDU5HT3RANCNFSM5O4JXPQA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
